pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+                   logPrefix, pam_strerror(pamh, result));
+        return PAM_IGNORE;//Since we are not an essential module, just make pam ignore us
+    }
+
+    struct passwd *userInfo;
+    userInfo = getpwnam(username);
+    if (!userInfo) {
+        pam_syslog(pamh, LOG_ERR, "%s: Couldn't get user info (passwd) info", logPrefix);
+        return PAM_IGNORE;
+    }
+
+    if (userInfo->pw_uid == 0) {
+        pam_syslog(pamh, LOG_DEBUG, "%s: Refusing to do anything for the root user", logPrefix);
+        return PAM_IGNORE;
+    }
+
+    const char *password;
+    result = pam_get_item(pamh, PAM_AUTHTOK, (const void**)&password);
+
+    if (result != PAM_SUCCESS) {
+        pam_syslog(pamh, LOG_ERR, "%s: Couldn't get password %s", logPrefix,
+                   pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    if (!password) {
+        pam_syslog(pamh, LOG_NOTICE, "%s: Couldn't get password (it is empty)", logPrefix);
+        //Asking for the password ourselves
+        result = prompt_for_password(pamh);
+        if (result != PAM_SUCCESS) {
+            pam_syslog(pamh, LOG_ERR, "%s: Prompt for password failed %s",
+                       logPrefix, pam_strerror(pamh, result)
+            );
+            return PAM_IGNORE;
+        }
+    }
+
+    //even though we just set it, better check to be 100% sure
+    result = pam_get_item(pamh, PAM_AUTHTOK, (const void**)&password);
+    if (result != PAM_SUCCESS || !password) {
+        pam_syslog(pamh, LOG_ERR, "%s: Password is not there even though we set it %s", logPrefix,
+                   pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    if (password[0] == '\0') {
+        pam_syslog(pamh, LOG_NOTICE, "%s: Empty or missing password, doing nothing", logPrefix);
+        return PAM_IGNORE;
+    }
+
+    char *key = strdup(password);
+    result = pam_set_data(pamh, kwalletPamDataKey, key, cleanup_free);
+
+    if (result != PAM_SUCCESS) {
+        free(key);
+        pam_syslog(pamh, LOG_ERR, "%s: Impossible to store the password: %s", logPrefix
+            , pam_strerror(pamh, result));
+        return PAM_IGNORE;
+    }
+
+    //TODO unlock kwallet that is already executed
+    return PAM_IGNORE;
+}
+
+static int drop_privileges(struct passwd *userInfo)
+{
+    /* When dropping privileges from root, the `setgroups` call will
+    * remove any extraneous groups. If we don't call this, then
+    * even though our uid has dropped, we may still have groups
+    * that enable us to do super-user things. This will fail if we
+    * aren't root, so don't bother checking the return value, this
+    * is just done as an optimistic privilege dropping function.
+    */
+    setgroups(0, NULL);
+
+    //Change to the user in case we are not it yet
+    if (setgid (userInfo->pw_gid) < 0 || setuid (userInfo->pw_uid) < 0 ||
+        setegid (userInfo->pw_gid) < 0 || seteuid (userInfo->pw_uid) < 0) {
```

Comment:
This could be done easier and more clearly using `setresuid()` and `setresgid()`